examples/ufo_toy.rs

```diff
@@ -0,0 +1,164 @@
+
+        for shape in self.glyph.outline.as_ref().iter().map(|o| o.contours.iter()).flatten() {
+            for point in shape.points.iter() {
+                println!("{:?}", point);
+                let color = if point.typ == PointType::OffCurve { &control_point_fg } else { &fg };
+                let circ = Circle::new((l_pad + (point.x * scale) as f64, (baseline - (point.y * scale) as f64)), 8.0 * scale as f64);
+                ctx.render_ctx.fill(circ, color, FillRule::NonZero);
+            }
+        }
+    }
+
+    fn layout(
+        &mut self,
+        bc: &BoxConstraints,
+        _children: &[Id],
+        _size: Option<(f32, f32)>,
+        _ctx: &mut LayoutCtx,
+    ) -> LayoutResult {
+        LayoutResult::Size(bc.constrain((100.0, 100.0)))
+    }
+}
+
+fn main() {
+    let glyph_path = match std::env::args().skip(1).next() {
+        Some(arg) => arg,
+        None => {
+            eprintln!("Please pass a path to a .glif file");
+            std::process::exit(1);
+        }
+    };
+
+    println!("loading {}", glyph_path);
+    let glyph = norad::Glyph::load(&glyph_path).expect("failed to load glyph");
+
+    druid_shell::init();
+
+    let mut run_loop = win_main::RunLoop::new();
+    let mut builder = WindowBuilder::new();
+    let mut state = UiState::new();
+
+    let root = GlyphEditor::new(glyph);
+    let root_id = state.add(root, &[]);
+    state.set_root(root_id);
+
+    builder.set_handler(Box::new(UiMain::new(state)));
+    builder.set_title("Ufo Toy");
+    let window = builder.build().expect("building window");
+
+    window.show();
+    run_loop.run();
+}
+
+fn make_path(contours: &[Contour]) -> BezPath {
+
+    /// An outline can have multiple contours, which correspond to subpaths
+    fn add_contour(path: &mut BezPath, contour: &Contour) {
+        let mut close: Option<&ContourPoint> = None;
+
+        if contour.points.is_empty() { return; }
+
+        let first = &contour.points[0];
+        path.moveto((first.x as f64, first.y as f64));
+        if first.typ != PointType::Move {
+            close = Some(first);
+        }
+
+        let mut idx = 1;
+        let mut controls: (Option<Vec2>, Option<Vec2>) = (None, None);
+
+        let mut add_curve = |to_point: Vec2, controls: &mut (Option<Vec2>, Option<Vec2>)| {
+            match (controls.0.take(), controls.1.take()) {
+                (Some(one), None) => path.quadto(one, to_point),
+                (Some(one), Some(two)) => path.curveto(one, two, to_point),
+                (None, None) => path.lineto(to_point),
+                _illegal => debug_assert!(false, "existence of second point implies first"),
+            }
+        };
+
+        while idx < contour.points.len() {
+            let next = &contour.points[idx];
+            let point: Vec2 = (next.x as f64, next.y as f64).into();
+            match next.typ {
+                PointType::OffCurve if controls.0.is_none() => controls.0 = Some(point.into()),
```

Comment:
I can't help but wonder if this might be cleaner for controls to just be a vec, of length 0-2.

Reply:
Quite possibly; in any case this will all get tossed in my next patch I think.